### PR TITLE
Make processed part of script read only when Coq is busy

### LINKED ISSUE #15675: CoqIDE frequently loses keyboard focus in 8.15
#### Description of the problem

Since Coq 8.15.0 CoqIDE frequently loses keyboard focus - at least on MacOS. One way to reproduce it is to create a new Window with `CMD+N`. Typing is then lost until one clicks into the Window. There are many other cases in which CoqIDE loses keyboard focus temporarily, that is after certain editing operations a few key strokes are lost, frequently the first one, but this is harder to reproduce. Still I have this at least 10x a per hour of CoqIDE usage. The described `Cmd+N` issue is 100% reproducible, though, in Coq 8.15 and does not exist in Coq 8.14.

I compared the Coq Platform 2022.01 versions of Coq 8.14 and 8.15.

I don't think that this is a Mac specific issue since 8.14 works fine, but can't exclude this. If there are doubts, I can test on other systems.

#### Coq Version

Broken:
```
coqtop -v
The Coq Proof Assistant, version 8.15.0
compiled with OCaml 4.10.2
```

Fine:
```
coqtop -v
The Coq Proof Assistant, version 8.14.1
compiled with OCaml 4.10.2
```

### LINKED ISSUE #15725: Can only delete accepted input in CoqIDE
#### Version

Coq 8.15.0

#### Operating system

Red Hat Enterprise Linux Workstation 7.9

#### Description of the problem

The following actions demonstrate an annoying regression,
where any attempt to modify accepted text results in its deletion.

* Open CoqIDE.
* Insert `Goal False.` into the buffer.
* Navigate forward.
* Select `False`.
* Start writing `True` in its place.

Observe how inserting `T` deletes `False` and
the subsequent inputs result in `Goal rue.`
becoming the contents of the buffer.

### LINKED ISSUE #15733: User modifications to processed (green) part of script while Coq is busy don't clear the "processed" highlight
Repeating an issue identified [here](https://github.com/coq/coq/pull/15714#issuecomment-1046089887) for clarity for users:

The green highlight indicates the parts of the script that have been processed and are valid.  If the user modifies processed text while Coq is busy processing an XML protocol request from CoqIDE, the buffer is modified without clearing the "processed" highlight even though it's now unprocessed and may be invalid.

For example, below I did a "Run to End", then I typed "GARBAGE", but the modified script is still shown in green even though it's now unprocessed and invalid.

Note that having active proof workers doesn't necessarily mean Coq is busy processing an XML request.

image

### PR DESCRIPTION
**Pull request by jfehrle**
In CoqIDE, prevents modifying _processed_ (green) parts of the script while Coq is busy (i.e. processing an XML request), which is important for the debugger.  Unlike the earlier code, it lets the user modify unmarked parts of the script (white/default background), which should address the main issue in #15675.  Note that code marked _to_process_ or _incomplete_ was already uneditable.

Includes some cleanup in `session.ml`.

@Alizter / @MSoegtropIMC  Perhaps you could try this in the next couple days?

I'd appreciate a prompt review so this can make it into 8.15.1.

Fixes: #15733
Fixes: #15675
Fixes: #15725

### CODE REVIEW

**Comment by Alizter** [comment]
So the issue I described in #15675 is definitely fixed by this patch. However I am unsure if @MSoegtropIMC's original issues are fixed.

I cannot reproduce window focus with Ctrl+n like @MSoegtropIMC described, before and after this patch, but I am on linux not mac.

Testing this I discovered another issue, which I've reported as #15715. Holding down Ctrl+n is pretty bad. You can hold down Ctrl+n for a few seconds and the memory usage skyrockets. It managed to brick my computer with 24gb of RAM, so I would be careful testing this. I will test if this is actually an issue present in the older versions as well.

**Comment by jfehrle** [comment]
Each buffer creates at least two processes (with exec rather than fork) so not surprising.

**Comment by jfehrle** [comment]
FWIW: "to brick": INFORMAL
cause (a smartphone or other electronic device) to become completely unable to function, **typically on a permanent basis**

--Mr Grammar Person

**Comment by ppedrot** [comment]
I think that this goes against the design of being able to modify Qed proofs in parallel without caring for what coq is performing.

**Comment by jfehrle** [comment]
As it frequently happens, there are multiple principles that are not easily made fully compatible.

Consider the green highlight, which indicates the parts of the script that have been processed and are valid.  The prior behavior breaks this, for example, at the bottom I did a "Run to End", then I typed "GARBAGE", but the modified script is still shown in green even though it's now unprocessed and invalid.

Two ways to address this case are:
- disable modifications in the processed area (as I've done in the PR).  This is the current behavior for the blue background text.
- abort execution and clear the green highlight after the point of the modification.  Current behavior when Coq is not running is to clear the highlight.

Aborting is undesirable because an inadvertent keypress could (surprisingly for the user) abort a minutes-long debugger session, which is likely disruptive.

image

**Comment by TheoWinterhalter** [comment]
Both options seem sensible to me, could it maybe maid configurable?

**Comment by ppedrot** [comment]
I think we should have a shortcut to start editing green areas with Qed proofs when coqc is performing unrelated type-checking. It should neither triggered by a random keypress, not be completely forbidden.

**Comment by jfehrle** [comment]
> Both options seem sensible to me, could it maybe made configurable?

Possible, of course, but making something configurable has costs, so the decision to add a new config option should be thought through.  In general terms, developers will need to implement, maintain and test both configuration settings.  Crudely speaking, twice the work.  Some configuration settings interact with others, making maintenance more difficult for all of the settings.

It's question of priorities.  Would it be better for users for me to first finish off the Ltac debugger (e.g. adding Ltac2 support) or to add the configuration option you suggest?

> I think we should have a shortcut to start editing green areas with Qed proofs when coqc is performing unrelated type-checking. It should neither triggered by a random keypress, not be completely forbidden.

That's a reasonable concept to consider as an enhancement.  The behavior needs to fleshed out (e.g. when/how is the modified script reprocessed?).  It may also be complex for the user.  Given that the highlighting code is event-driven, it would take some careful attention to implement and test.  I don't see this as a high priority; I think there are a number of other improvements to CoqIDE that would provide more value.

Do you know how PG or vsCoq handle this case?

The immediate question is what to do now for 8.15.1:
1. Nothing.  With the current code, the entire script is not editable at all while Coq is busy.  I expect you don't want that.
2. Make the read-only behavior the same as in 8.14.  Leave the inconsistencies for the indefinite future.  Not so great for debugger users.
3. Accept this PR as a bug fix and consider further improvements later on.

**Comment by jfehrle** [comment]
Note this comment from @gares at https://github.com/coq/coq/issues/15675#issuecomment-1046763860

> The other, a bit more complex, is to stop the computation: send an interrupt, maybe hide the "User interrupt" message, then clear highlights and go back to an earlier Stm state.

This one is very unreliable, so please [ignore insertions in processed code when Coq is busy].

**Comment by ejgallego** [comment]
The general consensus among UI people is that locking the editing buffer is a bad design pattern, which usually arise as the wrong solution to protocol deficiencies.

I won't say NACK on this as coqide or the protocol are beyond what I have time to contribute, but definitively this is going the wrong way.

**Comment by gares** [comment]
So what do you propose to implement as an immediate fix?

**Comment by Alizter** [comment]
The reproducible issue @MSoegtropIMC submitted in https://github.com/coq/coq/issues/15675#issuecomment-1047819351 and #15725 are not fixed by this patch. Specifically:
```coq
Goal True.
```
Stepping down, double clicking True and typing always discards the first key press.

**Comment by jfehrle** [comment]
> The general consensus among UI people is that locking the editing buffer is a bad design pattern, which usually arise as the wrong solution to protocol deficiencies.

Making part of the buffer read only in this case is not my first choice, but it seems the most consistent with what's already been done.  Avoiding it entirely would mean significant changes to both code and user experience.  For debuggers, updating the source code while you're debugging it has always been a challenging use case, generally leading to compromises/limitations.

**Comment by ppedrot** [comment]
> For debuggers, updating the source code while you're debugging

I don't see why a specific use case that is not the standard interaction mode should impact all users, regardless of the actual status of the debugger.

**Comment by jfehrle** [comment]
As I [pointed out](https://github.com/coq/coq/pull/15714#issuecomment-1046089887), it's not just the debugger.

**Comment by ejgallego** [comment]
> As I [pointed out](https://github.com/coq/coq/pull/15714#issuecomment-1046089887), it's not just the debugger.

This case is handled fine by main UI protocols, such as the ones in Isabelle or LSP [tho Isabelle has a bit of special setup for the sync, but that's due to other constraints that are not relevant here, and the plan was to remove the doc sync step which was a workaround for old bugs in their protocol IIUC]

**Comment by jfehrle** [comment]
Does anyone know what PG or vsCoq does in this case (in the screen shot)?

**Comment by ppedrot** [comment]
> it's not just the debugger.

I am not talking about a particular bug, but the design choices of the UI.

**Comment by jfehrle** [comment]
> The reproducible issue @MSoegtropIMC submitted in https://github.com/coq/coq/issues/15675#issuecomment-1047819351 and #15725 are not fixed by this patch.

@Alizter Please try it now.

**Comment by gares** [comment]
CoqIDE is not using LSP or PIDE, unfortunately. IMO the usability issue with this simple approach is that the "dropped input" should be signaled to the user, a bit like what the terminal does when there is not completion (it beeps, or blinks).

Once we have decent LSP support in Coq, then CoqIDE will have the chance to adopt it and get rid of this locking.
IMO it is much worse to end up with an inconsistent state (a broken text colored in green) than having CoqIDE reject your edit.

**Comment by SkySkimmer** [comment]
PG aborts and retracts (although it's a bit buggy as the edit ends up at the beginning of the aborted command, but the retract goes to before the place that the user was trying to edit)

**Comment by jfehrle** [comment]
For the future, perhaps Coq/IDEs could mark as unproven only those items that depend on earlier theorems that were modified (and are still unproven).  But not sure how hard it would be to get those dependencies correct and have the GUI always show a consistent state.  Nor whether the user benefit would justify the effort and whether such behavior would simple to use/understand.

**Comment by Alizter** [comment]
@jfehrle Seems to be working now.

The whole business of editing Qed proofs has always been buggy IIRC. There are even some nasty issues with the undo buffer getting corrupted leading to the entire undo operation not working properly.

**Review by gares — APPROVED** [review]

**Comment by jfehrle** [comment]
@ppedrot I added the `Ideutils.flash_info` call as you suggested, although what I had in mind was something more attention getting, such as changing the background color of the buffer for a fraction of a second.  I suppose you would think that's too intrusive.

@MSoegtropIMC Would you like to try this fix before 8.15.1?

**Comment by MSoegtropIMC** [comment]
> @MSoegtropIMC Would you like to try this fix before 8.15.1?

yes. Would you like me to test it before merge to main or after merge?

**Comment by jfehrle** [comment]
Perhaps test it before it's merged.  That might save creating a second PR if you find there's still an issue.  At the moment I don't plan to make further changes.

File: doc/changelog/09-coqide/15714-lost_keystrokes.rst
Comment (gares) [inline doc/changelog/09-coqide/15714-lost_keystrokes.rst:3]:
```suggestion
  Coq is busy evaluating a command are now ignored to
```
It is not about the wording, it is just that I don't expect users to even know that the protocol is XML based (and the changelog is for them)

Comment (jfehrle) [reply doc/changelog/09-coqide/15714-lost_keystrokes.rst:3]: Good point.  Made it "Coq is busy processing a request are now ignored to"

Can also be tactics as well as things like "setoption".

Comment (Alizter) [reply doc/changelog/09-coqide/15714-lost_keystrokes.rst]: Coq or CoqIDE though? Perhaps it doesn't matter since it's in the CoqIDE section.

Comment (jfehrle) [reply doc/changelog/09-coqide/15714-lost_keystrokes.rst]: I think "Coq" is most accurate.  CoqIDE is the obvious intermediary to talk to Coq.

**Comment by jfehrle** [comment]
@MSoegtropIMC Did you have a chance to try this fix?

**Comment by MSoegtropIMC** [comment]
@jfehrle : sorry not yet - too many other loose threads.

**Comment by jfehrle** [comment]
@MSoegtropIMC Thanks for checking.

@ppedrot Would you care to be the reviewer/assignee?

**Review by Alizter — APPROVED** [review]

**Comment by eponier** [comment]
What I observed while testing:
- the message "Input discarded" is sometimes shown without reason (I think one still needs to make it print at least once, but then it reappears sometimes, I don't understand why)
- in my setup, I have a really strange behaviour : I can enter as many "a" as I want in the green zone when it should be frozen; this is the only key on my keyboard that allows that
- the undo/redo has the right to change the green part

But I don't want to block that PR, because using coqide 8.15 is really painful at the moment wrt character inputs.

**Comment by Alizter** [comment]
@coqbot merge now

**Comment by jfehrle** [comment]
Too soon, I was fixing @eponier's issue.

**Comment by Alizter** [comment]
@jfehrle I announced my intention to merge a day in advance and @eponier has said they don't want to block this PR. How am I supposed to know that you are still working on this PR?

**Comment by jfehrle** [comment]
Well, if someone points out a problem with my PR, it would be good to give me a day to respond before merging (unless, perhaps, 8.15.1 will be frozen tomorrow).  These problems are not difficult to fix, which @eponier didn't know.  No big deal, I'll create another PR.

**Comment by Alizter** [comment]
OK fair enough. I will try to be less trigger happy in the future. You can also use labels like "needs: progress", emojis or something to communicate that you don't want the merge to go ahead.

**Pull request merged**
